pkg/sessions/dynamodb/dynamodb_store.go

```diff
@@ -0,0 +1,139 @@
+package dynamodb
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	dynamo "github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/sessions"
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/sessions/persistence"
+)
+
+const SessionKeyName = "SessionKey"
+
+// GetPutDeleteWithContext is a subset of the dynamodbiface.DynamoDBAPI interface that defines only the operations
+// we need. This is useful for testing as it enables us to provide a custom client that doesn't have to implement the full interface.
+type GetPutDeleteWithContext interface {
+	GetItemWithContext(aws.Context, *dynamo.GetItemInput, ...request.Option) (*dynamo.GetItemOutput, error)
+	PutItemWithContext(aws.Context, *dynamo.PutItemInput, ...request.Option) (*dynamo.PutItemOutput, error)
+	DeleteItemWithContext(aws.Context, *dynamo.DeleteItemInput, ...request.Option) (*dynamo.DeleteItemOutput, error)
+}
+
+// SessionStore is an implementation of the persistence.Store
+// interface that stores sessions in dynamoDB
+type SessionStore struct {
+	dynamoService GetPutDeleteWithContext
+	tableName     string
+}
+
+type DynamoSessionItem struct {
+	SessionKey string
+	Value      []byte
+	Expiry     string
+}
+
+var now = time.Now
+
+func NewDynamoDBSessionStore(opts *options.SessionOptions, cookieOpts *options.Cookie) (sessions.SessionStore, error) {
+	sess, err := session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("unable to connect to dynamodb: %v", err)
+	}
+
+	return persistence.NewManager(newSessionStore(dynamo.New(sess), opts.DynamoDB.TableName), cookieOpts), nil
+}
+
+func newSessionStore(client GetPutDeleteWithContext, tableName string) persistence.Store {
+	ss := &SessionStore{
+		dynamoService: client,
+		tableName:     tableName,
+	}
+
+	return ss
+}
+
+func (store *SessionStore) Save(ctx context.Context, key string, bytes []byte, duration time.Duration) error {
```

Comment:
NIT: I can see it is `[]byte` by the type, so the name `bytes` is redundant.  Might be more clear as a variable name like `value` to align with its purpose in the key/value use case in this method.